### PR TITLE
feat: EOF (Ethereum Object Format)

### PR DESCRIPTION
**Pull request by rakita**
Spec is here: https://github.com/ipsilon/eof/blob/main/spec/eof.md

List of EIPs that were merged in EOF, they are good to track progress:

- [x] 📃[EIP-3540](https://eips.ethereum.org/EIPS/eip-3540): EOF - EVM Object Format v1 [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-3540.md)
  - [x] EOF Decoding
  - [x] EOF Encoding 
- [x] 📃[EIP-3670](https://eips.ethereum.org/EIPS/eip-3670): EOF - Code Validation [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-3670.md)
- [x] 📃[EIP-4200](https://eips.ethereum.org/EIPS/eip-4200): EOF - Static relative jumps [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-4200.md)
- [x] 📃[EIP-4750](https://eips.ethereum.org/EIPS/eip-4750): EOF - Functions [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-4750.md)
- [ ] 📃[EIP-5450](https://eips.ethereum.org/EIPS/eip-5450): EOF - Stack Validation [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-5450.md)
- [x] 📃[EIP-6206](https://eips.ethereum.org/EIPS/eip-6206): EOF - JUMPF instruction [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-6026.md)
- [x] 📃[EIP-7480](https://eips.ethereum.org/EIPS/eip-7480): EOF - Data section access instructions [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-7480.md)
- [x] 📃[EIP-663](https://eips.ethereum.org/EIPS/eip-663): Unlimited SWAP and DUP instructions [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-663.md)
- [ ] 📃[EIP-7069](https://eips.ethereum.org/EIPS/eip-7069): Revamped CALL instructions (does not require EOF) [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-7069.md)
- [x] 📃[EIP-7620](https://eips.ethereum.org/EIPS/eip-7620): EOF - Contract Creation Instructions [history](https://github.com/ethereum/EIPs/commits/master/EIPS/eip-7620.md)

### CODE REVIEW

File: crates/interpreter/src/instructions/opcode.rs [hunk 1]
```diff
@@ -322,34 +324,34 @@ opcodes! {
...
+    0xE6 => DUPN => stack::dupn,
+    0xE7 => SWAPN => stack::swapn,
+    0xE8 => EXCHANGE => stack::exchange,
     // 0xE9
     // 0xEA
     // 0xEB
-    // 0xEC
-    // 0xED
-    // 0xEE
+    0xEC => EOFCREATE => contract::eofcrate::<H>,
```
Comment (DaniPopes) [inline crates/interpreter/src/instructions/opcode.rs:339]: is this not `CREATE3`? https://eips.ethereum.org/EIPS/eip-7620

File: crates/interpreter/src/instructions/utility.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+pub(crate) fn read_i16(ptr: *const u8) -> i16 {
```
Comment (DaniPopes) [inline crates/interpreter/src/instructions/utility.rs:1]: Should probably be `unsafe fn`

File: crates/interpreter/src/instructions/contract.rs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+use crate::{
+    gas::{self, cost_per_word, BASE, EOF_CREATE_GAS, KECCAK256WORD},
+    interpreter::{Interpreter, InterpreterAction},
+    primitives::{Address, Bytes, Eof, Spec, SpecId::*, B256, U256},
+    CallContext, CallInputs, CallScheme, CreateInputs, CreateScheme, EofCreateInput, Host,
+    InstructionResult, Transfer, MAX_INITCODE_SIZE,
+};
+use std::boxed::Box;
+
+pub fn eofcrate<H: Host>(interpreter: &mut Interpreter, host: &mut H) {
```
Comment (DaniPopes) [inline crates/interpreter/src/instructions/contract.rs:14]:
```suggestion
pub fn eofcreate<H: Host>(interpreter: &mut Interpreter, host: &mut H) {
```

**Comment by rakita** [comment]
@DaniPopes this PR is in very early draft, and opened only for progress tracking purposes

File: crates/interpreter/src/instructions/opcode.rs [hunk 4]
```diff
@@ -399,584 +605,120 @@ opcodes! {
...
-    while i < 256 {
-        table[i] = opcode_gas_info(i as u8, spec);
-        i += 1;
-    }
-    table
-}
-
-/// Returns a lookup table of opcode gas info for the given [`SpecId`].
-#[inline]
-pub const fn spec_opcode_gas(spec_id: SpecId) -> &'static [OpInfo; 256] {
```
Comment (DaniPopes) [inline crates/interpreter/src/instructions/opcode.rs:927]: How does this map onto the new opcode info? We need a way to get the static gas of opcodes for a specific SpecId

Comment (rakita) [reply crates/interpreter/src/instructions/opcode.rs:927]: I checked foundry, and it is only used to check if it is push opcode or not

**Comment by rakita** [comment]
This PR refactors a good amount of code related to OpCodes, Call inputs and Bytecode (Checked bytecode is removed, raw, analysed and EOF are now present. LockedBytecode is removed from Interpreter).

EOF is still not considered finished, i still need to verify and double check logic, call returns, stack requirements for CALLF and new EIP for future address expansion (https://github.com/ethereum/EIPs/pull/8385).

Verification for call or initcode types of bytecode still needs to be done, although this was discussed in last EOF call on best way forward. PR is passing all (but one) verification tests that evmone made, there are a lot of them and they cover a lot of use cases.

The main thing is that without EOF code deployment, EOF stays unaccessible, and ethereum/tests confirm the statement that all instructions are throwing errors if they are called outside of EOF code. And performance is similar.

So I am okay to merge it before I go on vacation and iterate on it when I am back.

**Pull request merged**